pkg/trace/operation_test.go

```diff
@@ -136,3 +137,173 @@ func TestSanity(t *testing.T) {
 		return
 	}
 }
+
+// MockHook is a testify mock that can be registered as a logrus hook
+type MockHook struct {
+	mock.Mock
+}
+
+// Levels indicates that the mock log hook supports all log levels
+func (m *MockHook) Levels() []logrus.Level {
+	return logrus.AllLevels
+}
+
+// Fire records that it has been called and returns an error if configured
+func (m *MockHook) Fire(entry *logrus.Entry) error {
+	args := m.Called(entry)
+
+	return args.Error(0)
+}
+
+// cases defines the set of messages we expect to see and the level we expect to see each at
+var cases = map[string]logrus.Level{
+	"DebugfMessage": logrus.DebugLevel,
+	"DebugMessage":  logrus.DebugLevel,
+	"InfofMessage":  logrus.InfoLevel,
+	"InfoMessage":   logrus.InfoLevel,
+	"WarnfMessage":  logrus.WarnLevel,
+	"WarnMessage":   logrus.WarnLevel,
+	"ErrorfMessage": logrus.ErrorLevel,
+	"ErrorMessage":  logrus.ErrorLevel,
+}
+
+// buildMatcher creates a testify MatchedBy function for the supplied operation
+func buildMatcher(op Operation, shouldContainOpId bool) func(entry *logrus.Entry) bool {
+	return func(entry *logrus.Entry) bool {
+		if shouldContainOpId && !strings.Contains(entry.Message, op.id) {
+			return false // Log message should have contained the operation id, but did not
+		}
+
+		if !shouldContainOpId && strings.Contains(entry.Message, op.id) {
+			return false // Log message should not have contained the operation id, but did
+		}
+
+
+		for message, level := range cases {
+			if entry.Level == level && strings.Contains(entry.Message, message) {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+// TestLogging demonstrates that log messages are relayed from the Operation to the Logger global
+func TestLogging(t *testing.T) {
+	original := Logger
+	defer func() { Logger = original }()
```

Comment:
This works but I normally pass the original value as an argument to the deferred function - it forces the value to bind at that time and protects against someone modifying `original` later in the function for whatever reason. No need for change in this PR.